src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -1183,6 +1183,13 @@ void IList.Clear()
             throw new NotSupportedException();
         }
 
+        private static bool IsCompatibleObject(object? value)
+        {
+            // Non-null values are fine.  Only accept nulls if T is a class or Nullable<U>.
+            // Note that default(T) is not equal to null for value types except when T is Nullable<U>.
+            return (value is T) || (value == null && default(T) == null);
```

Comment:
`default(T) == null` is treated as a constant by the JIT so the RHS of the expression can be optimized away if you express it like this:
```suggestion
            return (value is T) || (default(T) == null && value == null);
```

Reply:
Sure, why not?